lib/Tools/GenericSwizzling.cpp

```diff
@@ -0,0 +1,327 @@
+    }
+  }
+
+  auto smemReps = LinearLayout({{kVec, smem.getBases().lookup(kVec)},
+                                {kBank, smem.getBases().lookup(kBank)},
+                                {kSegment, unflatten(to_vector(segment))},
+                                {kReps, unflatten(to_vector(reps))}},
+                               smem.getOutDims(),
+                               /*requireSurjective=*/true);
+  return smemReps;
+  // if (reps.size() == 0) {
+  //   return smemReps;
+  // }
+  //// We now transpose the out dims so that reps are the last ones
+  //// In other words, we want the vec+bank+segment to be contiguous in smem
+  // auto repsMask = llvm::accumulate(reps, ~0u, [](uint32_t a, uint32_t b) {
+  // return a & b; }); auto rank = smemReps.getTotalOutDimSizeLog2();
+  //// Reshape to out dims = 2x2x...x2
+  // SmallVector<std::pair<StringAttr, int32_t>> splitDims;
+  // for (int32_t i = 0; i < rank; ++i) {
+  //   splitDims.push_back({StringAttr::get(ctx, std::to_string(i)), 2});
+  // }
+  // auto smemReps2 = smemReps.reshapeOuts(splitDims);
+
+  // SmallVector<StringAttr> front;
+  // SmallVector<StringAttr> back;
+  // for (int32_t i = 0; i < rank; ++i) {
+  //   if (repsMask & (1 << i)) {
+  //     back.push_back(StringAttr::get(ctx, std::to_string(i)));
+  //   } else {
+  //     front.push_back(StringAttr::get(ctx, std::to_string(i)));
+  //   }
+  // }
+  // auto transDims = llvm::to_vector(llvm::concat<StringAttr>(front, back));
+  // smemReps2 = smemReps2.transposeOuts(transDims);
+  // auto ret = smemReps2.reshapeOuts(smemReps.getOutDims());
+  //// we now have all the reps to be the trailing bases of the smem
+  // return ret;
+}
+
+SmallVector<int32_t> computeSegment(const SmallVector<int32_t> &bankSrc,
+                                    const SmallVector<int32_t> &bankDst,
+                                    int32_t rank, int32_t lenSegment) {
+  llvm::SmallDenseSet<int32_t> setSrc(bankSrc.begin(), bankSrc.end());
+  llvm::SmallDenseSet<int32_t> setDst(bankDst.begin(), bankDst.end());
+  // Remove the 0 as it's not a basis
+  setSrc.erase(0);
+  setDst.erase(0);
+
+  SmallVector<int32_t> segment;
+  for (int32_t b = 0; b < rank; ++b)
+    if (!setSrc.contains(1 << b) && !setDst.contains(1 << b))
+      segment.push_back(1 << b); // free variables
+  if (segment.size() >= lenSegment) {
+    segment.resize(lenSegment);
+    return segment;
+  }
+
+  // A and B are the difference sets
+  SmallVector<int32_t> A, B;
+  for (int32_t v : setSrc)
+    if (!setDst.contains(v))
+      A.push_back(v);
+  for (int32_t v : setDst)
+    if (!setSrc.contains(v))
+      B.push_back(v);
+  if (A.size() > B.size()) {
+    std::swap(A, B);
+  }
+  llvm::sort(A);
+  llvm::sort(B);
+  // A is the smaller set now
+  auto logBankConflicts = std::min<int32_t>(
+      std::max<int32_t>(0, lenSegment - A.size() - segment.size()), A.size());
+  // Conflict-free
+  for (int i = logBankConflicts; i < A.size(); ++i)
```

Comment:
This `^` operator here isn't clear to me, but we can chat offline

Reply:
this part is in the explanation of the algorithm in the paper, but yes, I agree that it is not super intuitive the first time you see it.

Reply:
Oh I reminded now this is the union of the two subspaces

Reply:
yep! This code finds the largest subspace that's not in the union of the two given ones.